pkg/engine/task/pod_exec.go

```diff
@@ -0,0 +1,201 @@
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"stat", "-c", "%s", file},
+		In:            nil,
+		Out:           &stdout,
+		Err:           nil,
+		TTY:           true, // this will forward 2>&1. otherwise, reading from Out will never return for e.g. missing files
+	}
+
+	if err := pe.Run(); err != nil {
+		return 0, fmt.Errorf("failed to get the size of %s, err: %v, stderr: %s", file, err, stdout.String())
+	}
+
+	raw := stdout.String()
+	trimmed := raw[:len(raw)-2] // remove trailing \n\r
+	size, err := strconv.ParseInt(trimmed, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse the size '%s' of the file %s : %v", raw, file, err)
+	}
+
+	return size, nil
+}
+
+// DownloadFile fetches a file from a remote pod. It runs `tar cf - file` command and streams contents
+// of the file via the stdout. Locally, the tar file is extracted into the passed afero filesystem where
+// it is saved under the same path. Afero filesystem is used to allow the caller downloading and persisting
+// of multiple files concurrently (afero filesystem is thread-safe).
+func DownloadFile(fs afero.Fs, file string, pod *v1.Pod, restCfg *rest.Config) error {
+	stdout := bytes.Buffer{}
+	stderr := strings.Builder{}
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"tar", "cf", "-", file},
+		In:            nil,
+		Out:           &stdout,
+		Err:           &stderr,
+	}
+	if err := pe.Run(); err != nil {
+		return fmt.Errorf("failed to copy pipe file. err: %v, stderr: %s", err, stderr.String())
+	}
+
+	if err := untarFile(fs, &stdout, file); err != nil {
+		return fmt.Errorf("failed to untar pipe file: %v", err)
+	}
+
+	return nil
+}
+
+// untarFile extracts a tar file from the passed reader using passed file name.
+func untarFile(fs afero.Fs, r io.Reader, fileName string) error {
+	tr := tar.NewReader(r)
+
+	for {
+		header, err := tr.Next()
+		if err != nil {
+			if err != io.EOF {
+				return err
+			}
+			break
+		}
+
+		// the target location of the file. tar strips the leading "/" however, we treat the pipe file path
+		// as a key to the underlying data (otherwise we'll have to start splitting paths). To avoid all
+		// the complexity and because we only extract one file here, the path is taken from the PipeFile configuration
+		target := fileName
+
+		// check the file type
+		switch header.Typeflag {
+		// if it's a file create it
+		case tar.TypeReg:
+			f, err := fs.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+			if err != nil {
+				return err
+			}
+
+			// copy over contents
+			if _, err := io.Copy(f, tr); err != nil {
+				return err
+			}
+
+			// manually close here after each file operation; deferring would cause each file close
+			// to wait until all operations have completed.
+			f.Close() // nolint
```

Comment:
If `io.Copy` fails, this leaks a file descriptor. How about adding scope to be able to defer here?
```go
case tar.TypeReg:
		{
			f, err := fs.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
			if err != nil {
				return err
			}
			defer f.Close()

			...
		}
```

Reply:
Good point